File: lib/logger.ts
type LogLevel = 'debug' | 'info' | 'warn' | 'error'

interface LogEntry {
  timestamp: string
  level: LogLevel
  message: string
  context?: Record<string, any>
  error?: Error
}

class Logger {
  private isDevelopment = process.env.NODE_ENV === 'development'
  private logLevel: LogLevel = (process.env.NEXT_PUBLIC_LOG_LEVEL as LogLevel) || 'info'

  private levelPriority: Record<LogLevel, number> = {
    debug: 0,
    info: 1,
    warn: 2,
    error: 3,
  }

  private shouldLog(level: LogLevel): boolean {
    return this.levelPriority[level] >= this.levelPriority[this.logLevel]
  }

  private formatLog(entry: LogEntry): string {
    const { timestamp, level, message, context, error } = entry
    const prefix = `[${timestamp}] [${level.toUpperCase()}]`

    let output = `${prefix} ${message}`

    if (context && Object.keys(context).length > 0) {
      output += ` ${JSON.stringify(context)}`
    }

    if (error) {
      output += `\nError: ${error.message}\nStack: ${error.stack}`
    }

    return output
  }

  private log(level: LogLevel, message: string, context?: Record<string, any>, error?: Error) {
    if (!this.shouldLog(level)) return

    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      level,
      message,
      context,
      error,
    }

    const formatted = this.formatLog(entry)

    // In development, use console with colors
    if (this.isDevelopment) {
      switch (level) {
        case 'debug':
          console.debug(formatted)
          break
        case 'info':
          console.info(formatted)
          break
        case 'warn':
          console.warn(formatted)
          break
        case 'error':
          console.error(formatted)
          break
      }
    } else {
      // In production, use standard console (which can be collected by logging services)
      console.log(formatted)
    }

    // In production, you could send logs to a service like:
    // - Sentry
    // - LogRocket
    // - Datadog
    // - CloudWatch
    // Example: if (level === 'error') sendToSentry(entry)
  }

  debug(message: string, context?: Record<string, any>) {
    this.log('debug', message, context)
  }

  info(message: string, context?: Record<string, any>) {
    this.log('info', message, context)
  }

  warn(message: string, context?: Record<string, any>) {
    this.log('warn', message, context)
  }

  error(message: string, error?: Error, context?: Record<string, any>) {
    this.log('error', message, context, error)
  }

  // API-specific logging helpers
  apiRequest(method: string, url: string, params?: Record<string, any>) {
    this.info(`API Request: ${method} ${url}`, params)
  }

  apiResponse(method: string, url: string, status: number, duration?: number) {
    this.info(`API Response: ${method} ${url}`, { status, duration })
  }

  apiError(method: string, url: string, error: Error) {
    this.error(`API Error: ${method} ${url}`, error)
  }

  // Feature-specific logging
  stockFetch(symbol: string, success: boolean) {
    if (success) {
      this.debug('Stock data fetched', { symbol })
    } else {
      this.warn('Stock data fetch failed', { symbol })
    }
  }

  portfolioAction(action: string, symbol?: string) {
    this.info(`Portfolio ${action}`, symbol ? { symbol } : undefined)
  }

  userAction(action: string, details?: Record<string, any>) {
    this.info(`User action: ${action}`, details)
  }
}

// Export singleton instance
export const logger = new Logger()

// Export for testing or custom instances
export { Logger }

File: backend/utils/logger.py
"""Logging utility for QuantPilot backend"""
import logging
import sys
from datetime import datetime
from typing import Any, Dict, Optional
from config import settings


class Logger:
    """Custom logger with structured logging support"""

    def __init__(self):
        self.logger = logging.getLogger("quantpilot")
        self.logger.setLevel(getattr(logging, settings.log_level.upper()))

        # Create console handler
        handler = logging.StreamHandler(sys.stdout)
        handler.setLevel(getattr(logging, settings.log_level.upper()))

        # Create formatter
        formatter = logging.Formatter(
            "[%(asctime)s] [%(levelname)s] %(message)s",
            datefmt="%Y-%m-%dT%H:%M:%S"
        )
        handler.setFormatter(formatter)

        # Add handler to logger
        if not self.logger.handlers:
            self.logger.addHandler(handler)

    def _format_context(self, context: Optional[Dict[str, Any]] = None) -> str:
        """Format context dictionary for logging"""
        if context and len(context) > 0:
            import json
            return f" {json.dumps(context)}"
        return ""

    def debug(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log debug message"""
        self.logger.debug(message + self._format_context(context))

    def info(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log info message"""
        self.logger.info(message + self._format_context(context))

    def warn(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log warning message"""
        self.logger.warning(message + self._format_context(context))

    def error(self, message: str, error: Optional[Exception] = None, context: Optional[Dict[str, Any]] = None):
        """Log error message"""
        error_msg = message + self._format_context(context)
        if error:
            self.logger.error(error_msg, exc_info=error)
        else:
            self.logger.error(error_msg)

    # API-specific logging helpers
    def api_request(self, method: str, url: str, params: Optional[Dict[str, Any]] = None):
        """Log API request"""
        self.info(f"API Request: {method} {url}", params)

    def api_response(self, method: str, url: str, status: int, duration: Optional[float] = None):
        """Log API response"""
        context = {"status": status}
        if duration:
            context["duration"] = duration
        self.info(f"API Response: {method} {url}", context)

    def api_error(self, method: str, url: str, error: Exception):
        """Log API error"""
        self.error(f"API Error: {method} {url}", error)

    # Feature-specific logging
    def stock_fetch(self, symbol: str, success: bool):
        """Log stock data fetch"""
        if success:
            self.debug("Stock data fetched", {"symbol": symbol})
        else:
            self.warn("Stock data fetch failed", {"symbol": symbol})

    def portfolio_action(self, action: str, symbol: Optional[str] = None):
        """Log portfolio action"""
        context = {"symbol": symbol} if symbol else None
        self.info(f"Portfolio {action}", context)

    def user_action(self, action: str, details: Optional[Dict[str, Any]] = None):
        """Log user action"""
        self.info(f"User action: {action}", details)


# Export singleton instance
logger = Logger()
